src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs

```diff
@@ -831,5 +832,138 @@ private void EmbedSourceLink(Stream stream)
+            {
+                portabilityPolicy |= identityComparer.PortabilityPolicy.SuppressSilverlightLibraryAssembliesPortability ? 0b1 : 0;
+                portabilityPolicy |= identityComparer.PortabilityPolicy.SuppressSilverlightPlatformAssembliesPortability ? 0b10 : 0;
+            }
+
+            WriteValue(CompilationOptionNames.PortabilityPolicy, portabilityPolicy.ToString());
+
+            var runtimeVersion = typeof(object).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+            WriteValue(CompilationOptionNames.RuntimeVersion, runtimeVersion);
+
+            var optimization = module.CommonCompilation.Options.OptimizationLevel == OptimizationLevel.Debug
+                ? (module.CommonCompilation.Options.DebugPlusMode ? "debug+" : "debug")
+                : "release";
+
+            WriteValue(CompilationOptionNames.Optimization, optimization);
+
+            module.CommonCompilation.SerializePdbEmbeddedCompilationOptions(builder);
+
+            _debugMetadataOpt.AddCustomDebugInformation(
+                parent: EntityHandle.ModuleDefinition,
+                kind: _debugMetadataOpt.GetOrAddGuid(PortableCustomDebugInfoKinds.CompilationOptions),
+                value: _debugMetadataOpt.GetOrAddBlob(builder));
+
+            void WriteValue(string key, string value)
+            {
+                builder.WriteUTF8(key);
+                builder.WriteByte(0);
+                builder.WriteUTF8(value);
+                builder.WriteByte(0);
+            }
+        }
+
+        /// <summary>
+        /// Writes information about metadata references to the pdb so the same
+        /// reference can be found on sourcelink to create the compilation again
+        /// </summary>
+        private void EmbedMetadataReferenceInformation(CommonPEModuleBuilder module)
+        {
+            var builder = new BlobBuilder();
+
+            // Order of information
+            // File name (null terminated string): A.exe
+            // Extern Alias (null terminated string): a1,a2,a3
+            // MetadataImageKind (byte)
+            // EmbedInteropTypes (boolean)
+            // COFF header Timestamp field (4 byte int)
+            // COFF header SizeOfImage field (4 byte int)
+            // MVID (Guid, 24 bytes)
+            foreach (var metadataReference in module.CommonCompilation.ExternalReferences)
+            {
+                if (metadataReference is PortableExecutableReference portableReference && portableReference.FilePath is object)
+                {
+                    var fileName = PathUtilities.GetFileName(portableReference.FilePath);
+                    var reference = module.CommonCompilation.GetAssemblyOrModuleSymbol(portableReference);
+                    var peReader = GetReader(reference);
+
+                    // Don't write before checking that we can get a peReader for the metadata reference
+                    if (peReader is null)
```

Comment:
What causes this case to happen? Is it deterministic?

Reply:
If it is not an `IAssemblySymbol` or `IModuleSymbol` we don't get a `PEReader ` for it

Reply:
Ah, I missed the local function. Makes sense.

---
In reply to: [436124491](https://github.com/dotnet/roslyn/pull/44373#discussion_r436124491) [](ancestors = 436124491)